release/releaser.go

```diff
@@ -45,539 +36,318 @@ func NewReleaser(
-	stage = metrics.NewTimer(base.With("stage", "fetch_platform_services"))
+	// From here in, we collect the results of the calculations.
+	results := flux.ReleaseResult{}
 
-	services, err := getServices.SelectServices(inst)
+	// Figure out the services involved.
+	logStatus("Finding defined services.")
+	timer = r.metrics.NewStageTimer("select_services")
+	var updates []*ServiceUpdate
+	updates, err = selectServices(rc, &spec, results, logStatus)
+	timer.ObserveDuration()
 	if err != nil {
-		return nil, errors.Wrap(err, "fetching platform services")
-	}
-	if len(services) == 0 {
-		res = append(res, r.releaseActionPrintf("No selected services found. Nothing to do."))
-		return res, nil
+		return nil, err
 	}
-
-	stage.ObserveDuration()
-	stage = metrics.NewTimer(base.With("stage", "calculate_applies"))
-
-	// Each service is running multiple images.
-	// Each image may need to be upgraded, and trigger an apply.
-	images, err := getImages.SelectImages(inst, services)
-	if err != nil {
-		return nil, errors.Wrap(err, "collecting available images to calculate applies")
+	logStatus("Found %d services.", len(updates))
+	report(results)
+
+	logStatus("Looking up images.")
+	if spec.ImageSpec != flux.ImageSpecNone {
+		timer = r.metrics.NewStageTimer("lookup_images")
+		// Figure out how the services are to be updated.
+		updates, err = calculateImageUpdates(rc.Instance, updates, &spec, results, logStatus)
+		timer.ObserveDuration()
+		if err != nil {
+			return nil, err
+		}
+		report(results)
 	}
 
-	updateMap := CalculateUpdates(services, images, func(format string, args ...interface{}) {
-		res = append(res, r.releaseActionPrintf(format, args...))
-	})
-
-	if len(updateMap) <= 0 {
-		res = append(res, r.releaseActionPrintf("All selected services are running the requested images. Nothing to do."))
-		return res, nil
+	// At this point we have have filtered the updates we can do down
+	// to nothing. Check and exit early if so
+	if len(updates) == 0 {
+		logStatus("No updates to do, finishing.")
+		report(results)
+		return nil, nil
 	}
 
-	stage.ObserveDuration()
-	stage = metrics.NewTimer(base.With("stage", "finalize"))
-
-	// We have identified at least 1 release that needs to occur. Releasing
-	// means cloning the repo, changing the resource file(s), committing and
-	// pushing, and then making the release(s) to the platform.
+	if spec.Kind == flux.ReleaseKindExecute {
+		if spec.ImageSpec != flux.ImageSpecNone {
+			logStatus("Pushing changes.")
+			timer = r.metrics.NewStageTimer("push_changes")
+			err = rc.PushChanges(updates, &spec)
+			timer.ObserveDuration()
+			if err != nil {
+				return nil, err
+			}
+		}
 
-	res = append(res, r.releaseActionClone())
-	for service, applies := range updateMap {
-		res = append(res, r.releaseActionUpdatePodController(service, applies))
+		logStatus("Applying changes.")
+		timer = r.metrics.NewStageTimer("apply_changes")
+		err = applyChanges(rc.Instance, updates, results)
```

Comment:
Oh, no it's not ignored. It is passed into `sendNotification(...)`, which sends a notification of whether the released completed or failed based on it, then spits it back.